test/test_models/pipeline-model/genai_config.json

```diff
@@ -0,0 +1,96 @@
+{
+    "model": {
+        "bos_token_id": 50256,
+        "context_length": 2048,
+        "decoder": {
+            "session_options": {
+                "log_id": "onnxruntime-genai",
+                "provider_options": [
+                    {}
+                ]
+            },
+            "head_size": 80,
+            "hidden_size": 2560,
+            "inputs": {
+                "input_ids": "input_ids",
+                "attention_mask": "attention_mask",
+                "past_key_names": "past_key_values.%d.key",
+                "past_value_names": "past_key_values.%d.value"
+            },
+            "outputs": {
+                "logits": "logits",
+                "present_key_names": "present.%d.key",
+                "present_value_names": "present.%d.value"
+            },
+            "num_attention_heads": 32,
+            "num_hidden_layers": 1,
+            "num_key_value_heads": 32,
+            "pipeline": [
+                {
+                    "embeddings": {
+                        "filename": "embeds.onnx",
+                        "inputs": [
+                            "input_ids"
+                        ],
+                        "outputs": [
+                            "/model/embed_tokens/Gather/output_0"
+                        ]
+                    },
+                    "transformer": {
+                        "filename": "transformer.onnx",
+                        "inputs": [
+                            "/model/embed_tokens/Gather/output_0",
+                            "attention_mask",
+                            "past_key_values.0.key",
+                            "past_key_values.0.value"
+                        ],
+                        "outputs": [
+                            "/model/layers.1/final_norm_layernorm/output_0",
+                            "present.0.key",
+                            "present.0.value"
+                        ],
+                        "session_options": {
```

Comment:
This doesn't have to be cleaned up now but having a `session_options` section in the `decoder` and then having another `session_options` section within the `transformer` section seems confusing. It may make more sense to have a `session_options` section for each `pipeline` entry (e.g. `embeddings`, `transformer`, `language_model_head`) and remove the global `session_options` section within `decoder`.

Reply:
The way it is currently structured is that the session options on the higher level is a fallback session options. In case users do not provide any session options on the pipeline model level, the fallback session options are used for that pipeline model session. The fallback session options also enables users to choose the device to use for search and sampling when the device used for pipelines is different compared to the one intended to be used for search and sampling.

Since the config structure is shared between the decoder-only models and the decoder-pipeline models, I cannot remove the session options on the decoder level. We could however change the structure to make it more user friendly, but that will break other published models.